### PR TITLE
add sphinx documentation (fix #197)

### LINKED ISSUE #197: Adding Sphinx documentation
# Demo

I have a not-guaranteed-to-be-up-to-date version of the documentation at http://hea-www.harvard.edu/~dburke/playground/sherpa/index.html - this should _not_ be taken as an official product as it is, in large part, an experiment. It may go away at any time!
# Background

As I know there's been discussion about this at [PyAstro16](http://python-in-astronomy.github.io/2016/), I thought it worth mentioning what I've learnt in my investigations to help anyone who wants to work on this. Please do!

There is a branch at [feature/sphinx-docs](/sherpa/sherpa/tree/feature/sphinx-docs/) which shows what I've done. I have a version on read-the-docs: [output](http://sherpa.readthedocs.org/en/latest/) and [configuration information](https://readthedocs.org/projects/sherpa/).

From memory, the issues include (they are _not_ ordered in severity, can overlap somewhat, and may actually feed into other possible improvements we could do):
- Move the contents of the existing `doc/` directory somewhere else (e.g. `notebook/`) so that `doc/` can be used for Sphinx.
- I had trouble getting Sherpa to build on Read The Docs as soon as I tried getting it to run code (e.g. to make sure the screen output is up to date); I forget the details but the [last log](https://readthedocs.org/projects/sherpa/builds/3261030/) suggests I gave up as soon as it complained about missing `yacc`. My thoughts at the time are that the build would have to be done locally, with the results pushed to the web site, but I didn't investigate this any further. There have been a number of build improvements to Sherpa since then, but we still need to compile code.
- I needed to add a customized IPython directive to grab the output from the Sherpa logger and include it into the Sphinx shell. It's pretty basic, but to save time re-inventing the wheel you can find it at  676cd6f77b5386d82065dd64439061bc1527a30a. If you don't do this then a lot of screen output is not included in the HTML output, which defeats the purpose just a little bit ;-)
  
  The version has since been updated (to address issues I found). It is less important when using the
  OO API, since most of the use of the logging infrastructure comes from the UI layer, but there are
  still places where it is used (e.g. the error analysis, although there I need to turn off the multi-cpu support for it to work properly, and I've yet to investigate why). [**note added Fri May  6 10:54:23 EDT 2016**]
- The presence of optional/multiple code paths is a challenge, both for building (what is included and what isn't) and actual documentation text: for example how to best explain functionality that may only be present in one I/O backend. This is a combination of issues: technological, documentation content, and social (what do we want).
- If you auto-generate the module layout then you end up with a confusing mess. Many Sherpa modules re-export content from other modules, which leads to a lot of repeated information and some very-large modules (e.g. `sherpa.astro.ui`), both of which lead to reader confusion and fatigue. If you don't autogenerate the layout, then you have a lot of work ahead of you, and you are still likely to hit some of these problems.
- There has been some clean up of the code since I worked on this, removing unused and CIAO-specific code, but there are probably still files left over that need to be excluded from the auto-generated documentation. Also, some of the modules are too large (I know I've already mentioned this, but it needs repeating).
- I am fairly sure some of the "clever" wrapping we do, to make some functionality easier to use from IPython, interacts poorly with Python docstrings, and hence the generated documentation. As an example, we allow users to say `sherpa.astro.ui.set_source(sherpa.astro.ui.polynom1d.cpt)` but `sherpa.astro.ui.polynom1d` is actually a `sherpa.ui.utils.ModelWrapper` instance around the `sherpa.models.basic.Polynom1D` class, which is where the docstrings should be (but currently are). Also, `Polynom1D` is also available as `sherpa.models.Polynom1D` (just re-iterating the point about repeated documentation). We (@olaurino and I) have talked about this particular issue before (i.e. pass through the model docstring) but that hasn't been worked on (and the use case of making sure that the information is available to Sphinx wasn't thought of).
  
  I keep on confusing myself about this; the problem is not for the object that was created, but for the class that creates the model object. That is, `help(sherpa.ui.polynom1d)` returns information on the `ModelWrapper` class. [**note clarified Wed May 11 12:02:38 EDT 2016**]
- I think we have good coverage of the UI API (although some is still missing and the content can be improved), except for a few specific areas such as the model instances and functions written in C, but the documentation of the lower-level routines is in a sorry state. Work on Sphinx documentation should help illuminate what areas need work, and can be done outside of this project (I have been trying to add stuff as I work on code fixes and improvements, in particular the C functions, but lots more to be done).
- See also my [comment on #22](https://github.com/sherpa/sherpa/issues/22#issuecomment-104009145) which discusses a problem I had with the DS9-related code when running Sphinx
  [**note added 16:25 EDT Friday 25 March 2016**]
- One suggestion that has been made is to not include the session-based API (what is referred to as the UI layer above - i.e. `sherpa.astro.ui` and `sherpa.ui`) in the Sphinx documentation. This would essentially have the UI layer documented as part of CIAO and the object-oriented API (which needs a lot of documentation) for the Sphinx documentation.
  [**note added 14:48 EDT Tuesday 26 April 2016**]

### PR DESCRIPTION
**Pull request by DougBurke**
This PR is ready for review, although I am likely to try and improve the documentation.

Summary
========

Create the Sherpa documentation using Sphinx to build, Travis-CI to test, and Read The Docs to build and display. The aim for the documentation is to build a sensible skeleton, but not to have a feature-complete set of documentation.

Notes
=====

There are three parts to getting sphinx documentation

  1. the infrastructure to get sphinx to build the documentation
  2. the contents of the documentation (mainly the .rst files but also fixes and additions to the
      module docstrings)
  3. getting docs built and available to the community

Status
=====

The outline of the documentation is present, as are much of the concepts, but there are missing areas and areas that need work.

All three points above have been addressed in this PR. I have set up a test build on Read The Docs - https://sherpa-test.readthedocs.io/en/latest/ - which requires manual re-building when new commits are made (so may not reflect the latest changes). Once the PR is integrated into the Sherpa branch I have a Read The Docs account set up to build the documentation from the master branch on every commit.
It is possible to use this PR but build the documentation in some other manner (e.g. via GitLab) as there are only a few small, text-only, configuration files specific to the Read The Docs environment:

  - `readthedocs.yml`
  - `docs/environment.yml`
  - `docs/rtd-pip-requirements`

In order to get the documentation to build on Read The Docs, which does not support "install any package you need", the build is done without requiring a functional Sherpa installation. It does this by mocking required packages *and* the compiled modules within Sherpa. The only required Python packages for the build (other than Sphinx and `sphinx_rtd_theme`) are `numpy` (since `setup.py` requires it), and `six` (although this probably also be mocked).

Previous issues with evidence of mocked symbols - e.g. in the routines provided in `sherpa.astro.xspec`, as described in https://sherpa-test.readthedocs.io/en/latest/model_classes/xspec_model.html - has been fixed.

The output from the example code is *not* auto-generated: this is different to my previous approach in  https://github.com/DougBurke/sherpa/tree/feature/sphinx-docs which was nice, but required the ability to build Sherpa *and* there was some **interesting** interaction between the Sherpa support code and some of the Sphinx utility routines with screen output.

Notes
=====

All the commits up to "`Add a Sphinx build`" are clean up, improvements, and moving information around in the existing code. It is not clear if all these changes are needed (since dropping support for building the documentation for Python 2.7 I have had less problems, but I have not gone tried the documentation build without these changes), but I believe they are worthwhile improvements anyway. They could also be cherry-picked into one or more PRs that are accepted before this one (getting these in sooner rather than later will reduce the amount of merge conflicts in our lives).

Read The Docs issues
=================

The main issue at the moment is that the build can time out, which has meant that I have removed the PDF and ePub generation to try and save time (they are also not useful at this moment). The runtime does seem somewhat random - presumably down to download times.

I was close to getting Read The Docs to be able to build Sherpa - which would have avoided the need to mock most things (although it may still be useful to mock out optional elements to reduce the round-trip time to get new documentation) by taking advantage of the conda support. Unfortunately there was a problem with the FORTRAN compiler (the conda environment didn't seem to be getting passed in), which scuppered these plans. There is a GitHub issue on the RTD page about this but I've not had any response.

### CODE REVIEW

**Comment by DougBurke** [comment]
Rebased. Tip was 604c495.

**Comment by DougBurke** [comment]
I've updated my local web site with the latest version - e.g. http://hea-www.harvard.edu/~dburke/playground/sherpa/models/index.html#available-models to see the model documentation, including the XSPEC models.

**Comment by DougBurke** [comment]
Rebased. Original tip was 0040d2f

**Comment by DougBurke** [comment]
Rebased. Original tip was 2b4cd9b

**Comment by olaurino** [comment]
Have you tried to run doctests on the sphinx documentation? I would expect it to fail at this point, because docstrings need special massaging in order to run with doctest, but it would be interesting to know how badly they fail.

**Comment by DougBurke** [comment]
I haven't tried doctest on this branch. I did play around with it in a previous attempt (hopefully it made it into the https://github.com/sherpa/sherpa/tree/feature/sphinx-docs branch but I don't have time to dig around there). At the moment I am not planning on investigating doctest, both from a technical side (there's enough issues I'm trying to get working already), and from the documentation side (not all code snippets make good or useful tests).

**Comment by DougBurke** [comment]
Rebased. Tip was 0cb1d92. Hopefully the `.travis.yml` changes aren't messed up here (since there was a conflict so they had to be manually merged).

**Comment by DougBurke** [comment]
I'm not 100% sold on the travis build set up, but it passes for now!

**Comment by DougBurke** [comment]
Rebased but I forgot to check what the tip was before rebasing. Oops.

**Comment by DougBurke** [comment]
Rebasing. Tip was a16a3f6

**Comment by DougBurke** [comment]
Rebasing. Tip was 009a4a3

**Comment by DougBurke** [comment]
About to rebase. Original tip was 9da4b6b

**Comment by DougBurke** [comment]
About to rebase. Tip was 9c6fa6e

**Comment by DougBurke** [comment]
About to rebase. Tip was c5ead39

**Comment by DougBurke** [comment]
About to rebase. Tip was a222c5c.

**Comment by DougBurke** [comment]
The XSPEC version support in PR #395 messes up the Sphinx doc build, since it hasn't mocked out the `sherpa.astro.xspec._xspec.get_xsversion` call, and so hilarity ensues. Need to look at when I get time.

**Comment by DougBurke** [comment]
About to rebase. Original tip was eb9db6a

**Comment by DougBurke** [comment]
Oops - I rebased to pull in the latest changes in `master` - and forgot to note down the previous head.

**Comment by DougBurke** [comment]
About rebase. Tip was 57491d8

**Comment by DougBurke** [comment]
About to rebase to pick up latest changes from master. Tip was efc9789.

**Comment by DougBurke** [comment]
About to rebase to try and consolidate the module changes (docstring, PEP8 fixes, tweaks to try and improve Sphinx builds) and separate them from the documentation changes.  This may not be the last-such rebase.

Tip was 068fddc

**Comment by DougBurke** [comment]
About to rebase: this is hopefully now split so that the changes to existing code (e.g. module docstrings and fixes for building the documentation using mock objects) are first, then the additions for the Sphinx documentation come second. I may try to combine some of the documentation changes (as having them as separate commits doesn't help anyone), but wanted to have this version saved first.

Tip was 080f03d

**Comment by DougBurke** [comment]
About to rebase. Some new documentation and some minor internal re-arrangement of commits.

Tip was 3672b0a.

**Comment by DougBurke** [comment]
About to rebase with a large simplification of the commit history.

Tip was 70500fe

**Comment by DougBurke** [comment]
About to rebase after merging virtually all of the Sphinx changes. Tip was 3672b0a.

File: helpers/extensions/__init__.py [hunk 1]
```diff
@@ -232,19 +235,45 @@ def build_lib_arrays(command, libname):
...
+#
+# It is not clear to DJB what exactly is going on under the hood
+# here, so this feels rather fragile, and system specific. Is there
+# a way to find out what the link flags would be without LDFLAGS
+# and then ensure they are included? At present this just assumes
+# that adding in -shared is enough.
+#
 ####
+
+f77_ldflags = os.environ.get('LDFLAGS', '')
```
Comment (olaurino) [inline helpers/extensions/__init__.py]: Before I forget, this seems like a hack, and it only solves one very specific problem on RTD. Other than that, the proper way of doing it is to change the relative option in `setup.cfg`, which is not a problem in a minimally decent build environment, as demonstrated by the fact that the CIAO build scripts, the Travis jobs configuration, the conda recipes being run on the Gitlab mirror or locally, etc. they all directly change the `setup.cfg` configuration for controlling the build.

**Comment by DougBurke** [comment]
About to rebase because of changes to master. Tip was 7cc9fe2.

**Comment by DougBurke** [comment]
@olaurino I rebased to pick up the latest changes and the OS-X test is failing to build, but as we don't see the log there's no way to know what went wrong - e.g. https://travis-ci.org/sherpa/sherpa/jobs/390111187

I'm tempted just to restart it in the hope that "turning it off and on again" will work, but do you have a better suggestion?

File: helpers/extensions/__init__.py (hunk 1 above)
Comment (DougBurke) [reply helpers/extensions/__init__.py]: Actually, it also "solves" the problem of the user who is trying to build the code with a recent conda package, without having to manually adjust the configuration. Or at least that is one of the use cases that I had for this commit.

Whether it's a good or workable solution is another matter, but I want to point out it's not just a Read-The-Docs issue.

**Comment by DougBurke** [comment]
About to rebase to fix `travis/setup_conda.sh` conflict. Tip was a533e3d

**Comment by DougBurke** [comment]
And on ReadTheDocs I get `Command killed due to excessive memory consumption`. Awesome sauce.

**Comment by DougBurke** [comment]
A test failure here is - I claim - nothing to do with this patch (unless it's how I handled rebasing the travis-build changes):

```
ImportError: dlopen(/Users/travis/build/sherpa/sherpa/sherpa/astro/xspec/_xspec.cpython-36m-darwin.so, 2): Library not loaded: @rpath/libtinfo.5.dylib
  Referenced from: /Users/travis/miniconda/envs/build/Xspec/x86_64-apple-darwin16.3.0/lib/libXSFunctions.dylib
  Reason: image not found
```

**Comment by DougBurke** [comment]
Rebasing to fix merge conflict in `sherpa/conftest.py`. Tip was cc920d4

**Comment by DougBurke** [comment]
I note that debian/ubuntu have packages for

```
saods9/bionic,bionic 7.5+repack1-2 all
  Image display tool for astronomy

xpa-tools/bionic,now 2.1.18-4 amd64 [installed]
  Tools for seamless communication between Unix programs
```

which we could install rather than grabbing a version from the DS9 site, but this doesn't help for the macOS build.

**Comment by DougBurke** [comment]
About to rebase to remove my DS9/Travis changes as this has now been fixed in master. Tip was bca9a2a (with travis changes) or beee312 (last change relevant to docs).

**Comment by DougBurke** [comment]
What do I need to do to get this PR accepted? Should I break up into a series of PRs to get the "non-doc changes" in first (since they are the ones that tend to cause merge conflicts)?

**Comment by DougBurke** [comment]
About to rebase. Tip was 2c5ed18

**Comment by DougBurke** [comment]
About to rebase to address test failure due to upstream commits. Tip was 63d011a

**Comment by DougBurke** [comment]
About to rebase to squash all the sphinx-related commits (and separate out a recent docstring change).

Original tip was 960a03a

**Comment by olaurino** [comment]
My main concert with this PR is that the code in the sphinx documentation cannot be tested. I think this is really a bad idea, as the benefits of testing the docs automatically greatly outweighs the complexity required to configure the tests. I guess I lost that battle, and that even if I didn't the argument would now be "it takes too long now", which is exactly why I advocating for enabling automated tests when we started talking about this. I'll merge the PR as is and I'll skip crossing my fingers as that would be futile.

**Comment by DougBurke** [comment]
We do not test any of the example code in the existing docstrings either. Shall we pull the whole code base until we fix this?

**Comment by olaurino** [comment]
No, that's not what I am saying. That's "legacy" docstrings, while the sphinx documentation is new and we could have enabled testing when we started if we wanted to.

**Comment by olaurino** [comment]
Rebasing to resolve conflicts, was fec0844. I'll wait for this to check out with Travis before merging anything else.

**Pull request merged**